addons/mrp/migrations/13.0.2.0/pre-migration.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2020 Payam Yasaie <https://www.tashilgostar.com>
+# Copyright 2020 Andrii Skrypka <andrijskrypa@ukr.net>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openupgradelib import openupgrade
+
+_model_renames = [
+    ('mrp.subproduct', 'mrp.bom.byproduct'),
+]
+
+_table_renames = [
+    ('mrp_subproduct', 'mrp_bom_byproduct'),
+    ('mrp_bom_line_product_attribute_value_rel', 'mrp_bom_line_product_template_attribute_value_rel'),
+]
+
+_field_renames = [
+    ('mrp.workorder', 'mrp_workorder', 'final_lot_id', 'finished_lot_id'),
+    ('mrp.bom', 'mrp_bom', 'sub_products', 'byproduct_ids'),
+    ('mrp.bom.line', 'mrp_bom_line', 'attribute_value_ids', 'bom_product_template_attribute_value_ids'),
+    ('mrp.production', 'mrp_production', 'propagate', 'propagate_cancel'),
+    ('stock.move', 'stock_move', 'subproduct_id', 'byproduct_id'),
+]
+
+_column_renames = {
+    'stock_move_line': [
+        ('lot_produced_id', None),
+    ],
+}
+
+_column_copies = {
+    'mrp_production': [
+        ('availability', 'reservation_state', None),
+    ],
+    'mrp_bom_line_product_template_attribute_value_rel': [
+        ('product_attribute_value_id', 'product_template_attribute_value_id', None),
+    ],
+}
+
+_xmlid_renames = [
+    # ir.model.access
+    ('mrp_byproduct.access_mrp_subproduct_manager', 'mrp.access_mrp_bom_byproduct_manager'),
+    ('mrp_byproduct.access_mrp_subproduct_user', 'mrp.access_mrp_bom_byproduct_user'),
+]
+
+
+def fill_bom_product_template_attribute_value(env):
+    """ Convert product.attribute.value ids to product.template.attribute.value ids"""
+    openupgrade.logged_query(env.cr, """
+    UPDATE mrp_bom_line_product_template_attribute_value_rel mbl_ptav_rel
+    SET product_template_attribute_value_id = (
+        SELECT id
+        FROM product_template_attribute_value ptav
+        WHERE mb.product_tmpl_id = ptav.product_tmpl_id AND mbl_ptav_rel.product_attribute_value_id = ptav.product_attribute_value_id
+        LIMIT 1
+    )
+    FROM mrp_bom_line mbl
+        JOIN mrp_bom mb ON mbl.bom_id = mb.id
+    WHERE mbl.id = mbl_ptav_rel.mrp_bom_line_id
+    """)
+
+
+def mapped_reservation_state(env):
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE mrp_production
+        SET reservation_state = CASE
+          WHEN reservation_state = 'partially_available' THEN 'assigned'
+          WHEN reservation_state = 'none' THEN NULL
+          END
+        WHERE reservation_state in ('partially_available', 'none')
+        """
+    )
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    openupgrade.remove_tables_fks(env.cr, 'mrp_bom_line_product_attribute_value_rel')
+    openupgrade.rename_models(env.cr, _model_renames)

```

Comment:
The code breaks here if `mrp_subproduct` was not installed in 12.0.
I suggest to add
```python
    if not openupgrade.table_exists(env.cr, 'mrp_subproduct'):
        create_model_table(env.cr, 'mrp_subproduct', 'Byproduct')
```
before this instruction, remembering to add an import for `create_model_table` to the top of the file:
`from odoo.tools import create_model_table`

Reply:
@pedrobaeza I think it has to been done in rename_tables of openupgradelib instead.
What do you think?

Reply:
@dennybiasiolli  Is OK now?

Reply:
It's perfect, thanks!